src/Stratis.Bitcoin.Features.SmartContracts/ReflectionExecutor/Controllers/BalancesController.cs

```diff
@@ -0,0 +1,184 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.Extensions.Logging;
+using NBitcoin;
+using Stratis.Bitcoin.Interfaces;
+using Stratis.SmartContracts.CLR;
+using Stratis.SmartContracts.Core.Util;
+
+namespace Stratis.Bitcoin.Features.SmartContracts.ReflectionExecutor.Controllers
+{
+    [ApiVersion("1")]
+    [Route("api/[controller]")]
+    public class BalancesController : Controller
+    {
+        private readonly ChainIndexer chainIndexer;
+        private readonly IBlockStore blockStore;
+        private readonly Network network;
+        private readonly ILogger logger;
+
+        public BalancesController(ChainIndexer chainIndexer,
+            IBlockStore blockStore,
+            Network network,
+            ILoggerFactory loggerFactory)
+        {
+            this.chainIndexer = chainIndexer;
+            this.blockStore = blockStore;
+            this.network = network;
+            this.logger = loggerFactory.CreateLogger(this.GetType().Name);
+        }
+
+        [Route("over-amount-at-height")]
+        [HttpGet]
+        public IActionResult GetBalancesOverAmount(int blockHeight, decimal amount)
+        {
+            Money thresholdBalance = Money.Coins(amount);
+
+            CoinviewContext coinView = GetCoinviewAtHeight(blockHeight);
+            Dictionary<Script, Money> balances = GetBalances(coinView);
+
+            List<Script> addressesWithThreshold = balances
+                .Where(x => x.Value >= thresholdBalance)
+                .Select(x => x.Key)
+                .ToList();
+
+            List<string> base58Addresses = new List<string>();
+
+            foreach (Script script in addressesWithThreshold)
+            {
+                GetSenderResult senderResult = this.GetAddressFromScript(script);
+
+                if (senderResult.Success)
+                {
+                    base58Addresses.Add(senderResult.Sender.ToBase58Address(this.network));
+                }
+                else
+                {
+                    this.logger.LogWarning($"{script} has a balance of {balances[script]} but is not a P2PK or P2PKH.");
+                }
+            }
+
+            return Json(base58Addresses);
+        }
+
+        private CoinviewContext GetCoinviewAtHeight(int blockHeight)
+        {
+            var coinView = new CoinviewContext();
+
+            const int batchSize = 1000;
+            IEnumerable<ChainedHeader> allBlockHeaders = this.chainIndexer.EnumerateToTip(this.chainIndexer.Genesis);
+
+            int totalBlocksCounted = 0;
+
+            int i = 0;
+
+            while (true)
```

Comment:
This logic is obvs VERY similar to the `UtxoIndexer` above. Was this intentional or can this be consolidated?